client/fetch_tagged_results_index_iterator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"github.com/m3db/m3db/serialize"
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3x/ident"
+)
+
+type fetchTaggedResultsIndexIterator struct {
+	currentIdx int
+	err        error
+	pools      fetchTaggedPools
+
+	current struct {
+		nsID ident.ID
+		tsID ident.ID
+		tags serialize.TagDecoder
+	}
+
+	backing struct {
+		nses [][]byte
+		ids  [][]byte
+		tags [][]byte
+	}
+}
+
+var _ index.Iterator = &fetchTaggedResultsIndexIterator{}
+
+func newFetchTaggedResultsIndexIterator(pools fetchTaggedPools) *fetchTaggedResultsIndexIterator {
```

Comment:
Take a look at how it's used in `fetchTaggedResultAccumulator.AsIndexQueryResults`: 

```
	var (
		iter  = newFetchTaggedResultsIndexIterator(pools)
		count = 0
	)
	results := fetchTaggedIDResultsSortedByID(accum.responses)
	sort.Sort(results)
	accum.responses = fetchTaggedIDResults(results)
	accum.responses.forEach(func(elems fetchTaggedIDResults) bool {
		iter.backing.ids = append(iter.backing.ids, elems[0].ID)
		iter.backing.nses = append(iter.backing.nses, elems[0].NameSpace)
		iter.backing.tags = append(iter.backing.tags, elems[0].EncodedTags)
		count++
		return count < limit
	})
```

Better to leave as is because it'll allow pooling of the underlying slices in `fetchTaggedResultsIndexIterator.backing.(nses|ids|tags)` once we start pooling `fetchTaggedResultsIndexIterator`.  Otherwise we'll have to manually pool the slices too.

Reply:
ok fair